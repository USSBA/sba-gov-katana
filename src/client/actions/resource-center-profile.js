import axios from 'axios'
import types from './types.js'
import constants from '../services/constants.js'

export function submitProfile(profile, honeyPotText) {
  return function(dispatch) {
    axios
      .post(constants.routes.submitProfile, {
        profile: profile,
        honeyPotText: honeyPotText
      })
      .then(response => {
        dispatch({
          type: types.resouceCenterProfile.submitComplete,
          payload: response.data,
          isSubmitComplete: true
        })
      })
      .catch(error => {
        dispatch({
          type: types.resouceCenterProfile.submitError,
          payload: error,
          hadSubmitError: true
        })
      })
  }
}
