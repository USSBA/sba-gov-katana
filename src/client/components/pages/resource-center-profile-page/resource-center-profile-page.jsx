import FormErrorMessage from '../../atoms/form-error-message/form-error-message'
import React from 'react'
import { CheckBox, Radio, Checkbox, MultiSelect, TextInput, LargePrimaryButton } from 'atoms'
import _ from 'lodash'
import {
  getPartners,
  getPartnerOffices,
  getPartnersByZip
} from '../../../services/resource-center-lookup.js'
import style from './resource-center-profile-page.scss'

class ResourceCenterProfilePage extends React.Component {
  constructor() {
    super()
    this.state = {
      loading: false,
      shouldUpdateContact: '',
      partners: getPartners(),
      offices: null,
      selectedOfficeOption: null,
      selectedOffice: null,
      isFieldValid: {
        type: null,
        name: null,
        address: null,
        phone: null,
        businessStage: null,
        serviceArea: null,
        url: null,
        hours: true, // optional so always valid
        expertise: null,
        services: null,
        languages: true // optional so always valid
      },

      profile: {
        type: null,
        name: null,
        address: null,
        phone: null,
        businessStage: '',
        serviceArea: null,
        url: '',
        hours: {
          mondayOpen: null,
          mondayClose: null,
          tuesdayOpen: null,
          tuesdayClose: null,
          wednesdayOpen: null,
          wednesdayClose: null,
          thursdayOpen: null,
          thursdayClose: null,
          fridayOpen: null,
          fridayClose: null,
          saturdayOpen: null,
          saturdayClose: null,
          sundayOpen: null,
          sundayClose: null
        },
        expertise: [],
        services: [],
        languages: []
      }
    }
  }

  onFocus() {}

  onBlur() {}

  updateOffices(partner) {
    this.setState({
      offices: _.reject(getPartnerOffices(partner), office => {
        return _.isEmpty(office)
      })
    })
    console.log(partner)
    console.log(getPartnerOffices(partner))
  }

  handleRadio(value, propName) {
    console.log(value)
    const newProfile = _.cloneDeep(this.state.profile)
    newProfile[propName] = value
    this.setState({ profile: newProfile })
  }

  handleCheckbox(event, propName) {
    const newProfile = _.cloneDeep(this.state.profile)
    const arrayToUpdate = newProfile[propName]
    const itemToUpdate = event.target.name
    if (event.target.checked) {
      arrayToUpdate.push(itemToUpdate)
    } else {
      _.remove(arrayToUpdate, item => {
        return item === itemToUpdate
      })
    }
    this.setState({ profile: newProfile })
  }

  handleSelect(newSelection, propName) {
    const value = newSelection.value
    const newProfile = _.cloneDeep(this.state.profile)
    console.log(propName)
    if (_.endsWith(propName, 'Open') || _.endsWith(propName, 'Close')) {
      newProfile.hours[propName] = value
    } else {
      newProfile[propName] = value
    }
    // if resource partner is changed
    if (propName === 'type') {
      this.updateOffices(newSelection.value)
      this.setState({
        profile: newProfile,
        selectedOffice: null,
        selectedOfficeOption: null
      })
    } else {
      this.setState({ profile: newProfile })
    }
  }

  handleChange(event, propName) {
    console.log(event)
    const newProfile = _.cloneDeep(this.state.profile)
    newProfile[event.target.name] = event.target.value
    this.setState({ profile: newProfile })
  }

  formatAddress(address) {
    console.log(address)
    if (!address) {
      return 'Please select your office'
    }
    const addressPartList = [address.street1, address.street2, address.city, address.state, address.zip]
    console.log(addressPartList)
    const addressString = _.join(
      _.filter(addressPartList, addressPart => {
        console.log(addressPart)
        console.log(!_.isEmpty(addressPart))
        return !_.isEmpty(addressPart)
      }),
      ', '
    )
    console.log(addressString)
    return addressString
  }

  handleOfficeSelect(newSelection) {
    console.log(newSelection)
    const newOffice = newSelection.value
    const newProfile = _.cloneDeep(this.state.profile)
    newProfile.name = newOffice.name1
    newProfile.phone = newOffice.phone
    newProfile.address = this.formatAddress(newOffice)
    this.setState({
      profile: newProfile,
      selectedOffice: newOffice,
      selectedOfficeOption: newSelection
      // shouldUpdateContact: this.state.shouldUpdateContact || !newProfile.phone || !newProfile.address
    })
  }

  validateFields() {
    const requiredFields = [
      'type',
      'name',
      'address',
      'phone',
      'businessStage',
      'serviceArea',
      'url',
      'expertise',
      'services'
    ]
    const newValidationState = _.cloneDeep(this.state.isFieldValid)
    _.forEach(requiredFields, field => {
      newValidationState[field] = !_.isEmpty(this.state.profile[field])
    })
    this.setState({
      isFieldValid: newValidationState,
      shouldUpdateContact:
        this.props.shouldUpdateContact || !newValidationState.address || !newValidationState.phone
    })
  }
  handleSubmit(event) {
    event.preventDefault()
    console.log('submit')
    console.log(this.state.profile)
    this.validateFields()
  }

  renderPartnerSelect() {
    const partners = _.map(this.state.partners, partner => {
      return {
        value: partner,
        label: partner
      }
    })

    return (
      <MultiSelect
        id="partner"
        value={this.state.profile.type}
        options={partners}
        multi={false}
        onChange={e => this.handleSelect(e, 'type')}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
      />
    )
  }

  renderOfficeSelect() {
    const offices = _.map(this.state.offices, office => {
      const locationArray = [office.city, office.state]
      let locationString = _.join(
        _.filter(locationArray, location => {
          return !_.isEmpty(location)
        }),
        ', '
      )
      if (locationString) {
        locationString = ' (' + locationString + ')'
      }
      return {
        value: office,
        label: office.name2 + locationString
      }
    })

    return (
      <MultiSelect
        id="office"
        value={this.state.selectedOfficeOption}
        options={offices}
        multi={false}
        onChange={this.handleOfficeSelect.bind(this)}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
      />
    )
  }
  renderAreaSelect() {
    const areas = _.map(['Entire Nation', 'Entire State', 'Entire City', 'Other'], area => {
      return {
        value: area,
        label: area
      }
    })

    return (
      <MultiSelect
        id="serviceArea"
        value={this.state.profile.serviceArea}
        options={areas}
        multi={false}
        onChange={e => this.handleSelect(e, 'serviceArea')}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
      />
    )
  }

  renderHourDropdowns() {
    const hours = _.map(
      [
        'Closed',
        '8:00 am',
        '8:30 am',
        '9:00 am',
        '9:30 am',
        '10:00 am',
        '10:30 am',
        '11:00 am',
        '11:30 am',
        '12:00 pm',
        '12:30 pm',
        '1:00 pm',
        '1:30 pm',
        '2:00 pm',
        '2:30 pm',
        '3:00 pm',
        '3:30 pm',
        '4:00 pm',
        '4:30 pm',
        '5:00 pm',
        '5:30 pm',
        '6:00 pm'
      ],
      hour => {
        return {
          value: hour,
          label: hour
        }
      }
    )
    const dayOfTheWeek = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
    const hoursOptions = _.map(dayOfTheWeek, day => {
      const dayOpen = day.toLowerCase() + 'Open'
      const dayClose = day.toLowerCase() + 'Close'
      return (
        <div className={style.hoursRow} key={day + '-container'}>
          <div className={style.hoursLabel}>
            <label>{day}</label>
          </div>
          <div className={style.hoursSelect}>
            <MultiSelect
              className={style.hoursSelect}
              id={dayOpen}
              value={this.state.profile.hours[dayOpen]}
              options={hours}
              multi={false}
              key={dayOpen}
              onChange={e => this.handleSelect(e, dayOpen)}
              onFocus={this.onFocus}
              onBlur={this.onBlur}
            />
          </div>
          <div className={style.hoursTo}>to</div>
          <div className={style.hoursSelect}>
            <MultiSelect
              className={style.hoursSelect}
              id={dayClose}
              value={this.state.profile.hours[dayClose]}
              options={hours}
              multi={false}
              key={dayClose}
              onChange={e => this.handleSelect(e, dayClose)}
              onBlur={this.onBlur}
              onFocus={this.onFocus}
            />
          </div>
        </div>
      )
    })
    return <div>{hoursOptions}</div>
  }

  renderExpertiseCheckboxes() {
    const expertiseOptions = _.map(
      [
        'Creating a plan',
        'Getting a loan',
        'Business finances',
        'Marketing and sales',
        'Government contracting',
        'Legal issues',
        'International trade',
        'Networking'
      ],
      expertise => {
        return (
          <Checkbox
            id={expertise}
            name={expertise}
            label={expertise}
            key={expertise}
            handleChange={e => this.handleCheckbox(e, 'expertise')}
            checked={this.state.profile.expertise.includes(expertise)}
          />
        )
      }
    )
    return expertiseOptions
  }
  renderServiceCheckboxes() {
    const serviceOptions = _.map(
      [
        'Walk-ins',
        'Appointments',
        'In-person',
        'By phone',
        'By video chat or online',
        'Trainings or workshops',
        'Certifications',
        'Night or weekend availability'
      ],
      service => {
        return (
          <Checkbox
            id={service}
            name={service}
            label={service}
            key={service}
            handleChange={e => this.handleCheckbox(e, 'services')}
            checked={this.state.profile.services.includes(service)}
          />
        )
      }
    )
    return serviceOptions
  }

  renderLanguageCheckboxes() {
    const languageOptions = _.map(
      [
        'Spanish',
        'Chinese',
        'Tagalog',
        'Vietnamese',
        'French',
        'German',
        'Korean',
        'Arabic',
        'Russian',
        'Other'
      ],
      language => {
        return (
          <Checkbox
            id={language}
            name={language}
            label={language}
            key={language}
            checked={this.state.profile.languages.includes(language)}
            handleChange={e => this.handleCheckbox(e, 'languages')}
          />
        )
      }
    )

    return languageOptions
  }

  renderBusinessStageRadios() {
    const businessStageOptions = [
      {
        value: 'New and early stage businesses',
        text: 'New and early stage businesses'
      },
      {
        value: 'Established businesses',
        text: 'Established businesses'
      }
    ]
    return (
      <Radio
        options={businessStageOptions}
        value={this.state.profile.businessStage}
        onChange={e => this.handleRadio(e, 'businessStage')}
        textStyle={style.radioText}
        onBlur={this.onBlur}
        onFocus={this.onFocus}
      />
    )
  }

  renderShouldUpdateAddressRadios() {
    const id = 'resource-center-profile-form'

    const updateContactInfoOptions = [
      {
        value: 'false',
        text: 'Yes'
      },
      {
        value: 'true',
        text: 'No, it needs to be updated'
      }
    ]
    return (
      <Radio
        id={id + '-should-update-contact-info'}
        //errorText={constants.messages.validation.invalidIndustryExperience}
        label=""
        name="shouldUpdateContactInfo"
        onChange={e => this.setState({ shouldUpdateContact: e === 'true' })}
        //validationState={this.state.validStates.industryExperience}
        value={this.state.shouldUpdateContact.toString()}
        options={updateContactInfoOptions}
        textStyle={style.radioText}
        onBlur={this.onBlur}
        onFocus={this.onFocus}
      />
    )
  }

  render() {
    const id = 'resource-center-profile-form'
    const selectedOffice = this.state.selectedOffice
    const isFieldValid = this.state.isFieldValid

    return (
      <div className={style.backgroundContainer}>
        <div className={style.container}>
          <form id={id} className={style.form}>
            <h2>Resource Center Profile</h2>
            <p>
              Answer the following questions about your office's expertise to be better matched with your
              ideal clients
            </p>
            <label className={isFieldValid.type === false ? style.invalid : ''}>
              Which resource partner do you represent?
            </label>
            {this.renderPartnerSelect()}
            {this.state.profile.type && (
              <label className={isFieldValid.name === false ? style.invalid : ''}>Which office?</label>
            )}
            {this.state.profile.type && this.renderOfficeSelect()}
            {selectedOffice && <label>Is this your office address and phone number?</label>}
            {selectedOffice && (
              <div>
                <div>{selectedOffice.street1}</div>
                <div>{selectedOffice.street2}</div>
                <div>
                  {selectedOffice.city ? selectedOffice.city + ', ' : ''}
                  {selectedOffice.state ? selectedOffice.state + ', ' : ''}
                  {selectedOffice.zip ? selectedOffice.zip : ''}
                </div>
                <div>{selectedOffice.phone}</div>
                {this.renderShouldUpdateAddressRadios()}
                {this.state.shouldUpdateContact && (
                  <div>
                    <label className={isFieldValid.address === false ? style.invalid : style.formLabel}>
                      What's your address?
                    </label>
                    <TextInput
                      id={id + '-address'}
                      name="address"
                      value={this.state.profile.address}
                      onChange={this.handleChange.bind(this)}
                      autoFocus={false}
                      onBlur={this.onBlur.bind(this)}
                      onFocus={this.onFocus.bind(this)}
                    />

                    <label className={isFieldValid.phone === false ? style.invalid : style.formLabel}>
                      What's your phone number?
                    </label>
                    <TextInput
                      id={id + '-phone'}
                      name="phone"
                      value={this.state.profile.phone}
                      onChange={this.handleChange.bind(this)}
                      autoFocus={false}
                      onBlur={this.onBlur.bind(this)}
                      onFocus={this.onFocus.bind(this)}
                    />
                  </div>
                )}
              </div>
            )}
            <label className={isFieldValid.url === false ? style.invalid : ''}>
              What's your website URL?
            </label>
            <TextInput
              id={id + '-website-url'}
              //errorText={constants.messages.validation.invalidName}

              name="websiteUrl"
              onChange={this.handleChange.bind(this)}
              value={this.state.profile.websiteUrl}
              //validationState={this.state.validStates.contactFullName}
              autoFocus={false}
              onBlur={this.onBlur.bind(this)}
              onFocus={this.onFocus.bind(this)}
            />
            <label>
              What are your normal office hours? <span className={style.optional}>(optional)</span>
            </label>
            <div className={style.grid}>{this.renderHourDropdowns()}</div>

            <label className={isFieldValid.serviceArea === false ? style.invalid : ''}>
              What area does your office serve?
            </label>
            {this.renderAreaSelect()}

            <label className={isFieldValid.services === false ? style.invalid : ''}>
              What services does your office provide?
            </label>
            {this.renderServiceCheckboxes()}

            <label className={isFieldValid.expertise === false ? style.invalid : ''}>
              In which three topic areas does your office have the most expertise?
            </label>
            {this.renderExpertiseCheckboxes()}

            <label className={isFieldValid.businessStage === false ? style.invalid : ''}>
              Which business stage does your office best serve?
            </label>
            {this.renderBusinessStageRadios()}

            <label>
              In which languages other than English can your office provide counseling{' '}
              <span className={style.optional}>(optional)</span>
            </label>
            {this.renderLanguageCheckboxes()}
            <div className={style.submitButton} onClick={this.handleSubmit.bind(this)}>
              <LargePrimaryButton id="feedback-submit-button" text="SUBMIT" url="" />
            </div>
            {_.some(isFieldValid, field => {
              console.log(field)
              return field === false
            }) && (
              <div className={style.submitButton}>
                <FormErrorMessage errorText="Please answer all required questions before submitting." />
              </div>
            )}
          </form>
        </div>
      </div>
    )
  }
}

export default ResourceCenterProfilePage
/***************
 Expected POST Body Sample:
 {
   "profile": {
     "type": "WBC",
     "name": "My WBC Center",
     "address": "8 Market Place, Baltimore, MD 21202",
     "phone": "123-456-7890",
     "businessStage": "Stage51",
     "serviceArea": "Area51",
     "url": "www.example.com",
     "hours": {
       "mondayOpen": "9:00 am",
       "mondayClose": "5:00 pm",
       "tuesdayOpen": "9:00 am",
       "tuesdayClose": "5:00 pm",
       "wednesdayOpen": "9:00 am",
       "wednesdayClose": "5:00 pm",
       "thursdayOpen": "9:00 am",
       "thursdayClose": "5:00 pm",
       "fridayOpen": "9:00 am",
       "fridayClose": "5:00 pm",
       "saturdayOpen": "9:00 am",
       "saturdayClose": "5:00 pm",
       "sundayOpen": "9:00 am",
       "sundayClose": "5:00 pm"
     },
     "expertise": [
       "Being Cool",
       "Being Rad"
     ],
     "services": [
       "Cool Consultancy",
       "Rad Consultancy"
     ],
     "languages": [
       "Skater",
       "Surfer"
     ]
   }
 }
 */
