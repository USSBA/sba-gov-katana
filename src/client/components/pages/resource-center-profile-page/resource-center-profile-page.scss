label,
.radioText {
  font-weight: normal;
  font-size: 1em !important;
}

.optional {
  font-weight: lighter;
}
.hoursRow {
  margin-bottom: 0.375em;
  @include grid-media($sba-grid-small) {
    height: 9em;
  }
  @include grid-media($sba-grid-medium) {
    height: 2.6em;
    margin-right: -8%;
  }
  @include grid-media($sba-grid-large) {
    height: 2.6em;
    margin-right: -12%;
  }
  line-height: 2.6em;
}
.hoursRow .Select {
  display: inline-block;
}

.hoursRow > div {
  display: inline-block;
  //max-width: 40%;
  // margin-left: 10px;
  // margin-right: 10px;
  vertical-align: middle;
  margin-bottom: 0.375em;
}
.hoursRow label {
  margin-bottom: 0px;
}
div.hoursLabel {
  //width: 3em !important;
  margin-bottom: 0px;
  @include grid-media($sba-grid-small) {
    @include grid-column(6);
  }
  @include grid-media($sba-grid-medium) {
    @include grid-column(3);
    margin-left: 0px !important;
  }
  @include grid-media($sba-grid-large) {
    @include grid-column(3);
    margin-left: 0px !important;
  }
}
div.hoursSelect {
  @include grid-media($sba-grid-small) {
    @include grid-column(5);
  }
  @include grid-media($sba-grid-medium) {
    @include grid-column(4);
  }
  @include grid-media($sba-grid-large) {
    @include grid-column(4);
  }
}
.hoursTo {
  @include grid-column(1);
  text-align: center;
  margin-bottom: 0px;
}
.hours {
  // display: grid;
  columns: 12;
}
.backgroundContainer {
  background: $sba-blue-lightest;
  margin: 0px 0px;
  padding: 20px 20px;
  @include grid-media($sba-grid-small) {
    @include grid-column(6, $mobile-0gutter);
  }
  @include grid-media($sba-grid-medium) {
    @include grid-column(12, $tablet-0gutter);
  }
  @include grid-media($sba-grid-large) {
    @include grid-column(12, $desktop-0gutter);
  }
}

.container {
  padding-left: 20px;
  padding-right: 20px;
  background: white;
  box-shadow: 2px 2px 15px #888888;
  @include grid-media($sba-grid-small) {
    @include grid-column(6);
    margin-top: 20px;
    margin-bottom: 40px;
    padding-top: 20px;
  }
  @include grid-media($sba-grid-medium) {
    @include grid-column(12);
    margin-top: 20px;
    margin-bottom: 30px;
    padding: 5%;
  }
  @include grid-media($sba-grid-large) {
    @include grid-column(8);
    @include grid-push(2);
    margin-top: 20px;
    margin-bottom: 30px;
    padding: 5%;
  }
  .maidenNameContainer {
    display: none;
  }

  form,
  .submissionMessage {
    @include grid-media($sba-grid-small) {
      max-width: 100%;
    }
    @include grid-media($sba-grid-medium) {
      max-width: 90%;
    }
    @include grid-media($sba-grid-large) {
      max-width: 75%;
    }
    margin: 0px auto;
  }

  .submitButton {
    margin: 20px auto;
    text-align: center;
  }

  form > label,
  label.formLabel {
    font-weight: bold;
    margin-top: 2em;
  }

  .invalid {
    font-weight: bold;
    margin-top: 2em;
    color: $error;
  }

  .invalid::before {
    content: '* ';
  }

  .hidden {
    display: none;
  }

  h1 {
    @include font-size($h3-font-sizes);
  }
  .submissionMessage > h1 {
    text-align: center;
  }
}
