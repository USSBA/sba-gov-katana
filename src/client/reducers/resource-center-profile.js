import types from '../actions/types.js'

const resourceCenterProfileReducer = (state = {}, action) => {
  if (action.type === types.resouceCenterProfile.submitComplete) {
    return { isSubmitComplete: action.isSubmitComplete }
  } else if (action.type === types.resouceCenterProfile.submitError) {
    return { hadSubmitError: action.hadSubmitError }
  }
  return state
}

export default resourceCenterProfileReducer
