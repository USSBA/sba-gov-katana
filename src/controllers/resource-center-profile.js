import { saveProfile, getProfiles } from "../service/resource-center-profile.js";
import _ from "lodash";
import HttpStatus from "http-status-codes";

/***************
 Expected POST Body Sample:
 {
   "profile": {
     "type": "WBC",
     "name": "My WBC Center",
     "address": "8 Market Place, Baltimore, MD 21202",
     "phone": "123-456-7890",
     "businessStage": "Stage51",
     "serviceArea": "Area51",
     "url": "www.example.com",
     "hours": {
       "mondayOpen": "9:00 am",
       "mondayClose": "5:00 pm",
       "tuesdayOpen": "9:00 am",
       "tuesdayClose": "5:00 pm",
       "wednesdayOpen": "9:00 am",
       "wednesdayClose": "5:00 pm",
       "thursdayOpen": "9:00 am",
       "thursdayClose": "5:00 pm",
       "fridayOpen": "9:00 am",
       "fridayClose": "5:00 pm",
       "saturdayOpen": "9:00 am",
       "saturdayClose": "5:00 pm",
       "sundayOpen": "9:00 am",
       "sundayClose": "5:00 pm"
     },
     "expertise": [
       "Being Cool",
       "Being Rad"
     ],
     "services": [
       "Cool Consultancy",
       "Rad Consultancy"
     ],
     "languages": [
       "Skater",
       "Surfer"
     ]
   }
 }
 */

function handleProfileSubmission(req, res) {
  if (_.has(req, "body.profile")) {
    const profile = _.assign(
      {
        sourceIpAddress: req.ip
      },
      req.body.profile
    );
    saveProfile(profile, req.body.honeyPotText)
      .then(function() {
        res.status(HttpStatus.NO_CONTENT).send();
      })
      .catch((error) => {
        console.error(error);
        res
          .status(HttpStatus.INTERNAL_SERVER_ERROR)
          .send("Error saving resource center profile.");
      });
  } else {
    res
      .status(HttpStatus.BAD_REQUEST)
      .send(
        "Profile missing from body.  Please check that the HTTP request includes a profile value"
      );
  }
}

function retrieveProfiles(req, res) {
  // There is currently no authorization required for requests to proceed
  if (req) {
    getProfiles(req.sessionInfo)
      .then(function(results) {
        res
          .header("Content-Type", "application/json")
          .status(HttpStatus.OK)
          .send(results);
      })
      .catch((error) => {
        if (error.message === "FORBIDDEN") {
          res
            .status(HttpStatus.FORBIDDEN)
            .send(
              "Please log in as an Administrator before requesting this data"
            );
        } else {
          console.log("Error Thrown during Profile Retrieval", error);
          res
            .status(HttpStatus.INTERNAL_SERVER_ERROR)
            .send("Error retrieving Resource Center Profiles");
        }
      });
  } else {
    res
      .status(HttpStatus.FORBIDDEN)
      .send("Please log in as an Administrator before requesting this data.");
  }
}

export { retrieveProfiles, handleProfileSubmission };
