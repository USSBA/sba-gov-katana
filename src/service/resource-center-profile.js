/* eslint-disable id-length,object-property-newline */
import * as ResourceCenterModel from '../models/resource-center-profile.js'
import _ from 'lodash'
import moment from 'moment'
import uuid from 'uuid'
import { isAdministrator } from '../service/user-service.js'
import { formatResourceCenterData } from '../util/formatter.js'

function saveProfile(profile, honeyPotText) {
  if (honeyPotText) {
    console.log('Detected submission with honeypot field filled. Dropping save.')
    return Promise.resolve()
  }
  const newProfile = _.assign({}, profile, {
    id: uuid.v4(),
    timestamp: moment().unix(),
    expertise: formatArray(profile.expertise, 'expertise'),
    services: formatArray(profile.services, 'service'),
    languages: formatArray(profile.languages, 'language')
  })

  return ResourceCenterModel.resourceCenterProfile.create(newProfile, {
    include: [
      {
        model: ResourceCenterModel.resourceCenterHours,
        as: 'hours'
      },
      {
        association: ResourceCenterModel.expertise,
        as: 'expertise'
      },
      {
        association: ResourceCenterModel.services,
        as: 'services'
      },
      {
        association: ResourceCenterModel.languages,
        as: 'languages'
      }
    ]
  })
}

function getProfiles(sessionId) {
  return ResourceCenterModel.resourceCenterProfile
    .findAll({
      order: [['timestamp', 'DESC']],
      attributes: ['name', 'address', 'type', 'timestamp', 'businessStage', 'serviceArea', 'url', 'phone'],
      include: [
        {
          model: ResourceCenterModel.resourceCenterHours,
          as: 'hours',
          attributes: [
            'mondayOpen',
            'mondayClose',
            'mondayOpen',
            'mondayClose',
            'tuesdayOpen',
            'tuesdayClose',
            'wednesdayOpen',
            'wednesdayClose',
            'thursdayOpen',
            'thursdayClose',
            'fridayOpen',
            'fridayClose',
            'saturdayOpen',
            'saturdayClose',
            'sundayOpen',
            'sundayClose'
          ]
        },
        {
          association: ResourceCenterModel.expertise,
          as: 'expertise',
          attributes: ['expertise']
        },
        {
          association: ResourceCenterModel.services,
          as: 'services',
          attributes: ['service']
        },
        {
          association: ResourceCenterModel.languages,
          as: 'languages',
          attributes: ['language']
        }
      ]
    })
    .then(data => {
      if (data && data.length > 0) {
        return formatResourceCenterData(data)
      }
      return ''
    })
}

function formatArray(elements, fieldName) {
  return _.map(elements, element => {
    return _.set({}, fieldName, element)
  })
}

export { saveProfile, getProfiles }
