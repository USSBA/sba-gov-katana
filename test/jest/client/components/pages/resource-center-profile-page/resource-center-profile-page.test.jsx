import React from 'react'
import renderer from 'react-test-renderer'
import _ from 'lodash'
import { shallow } from 'enzyme'

import { ResourceCenterProfilePage } from 'pages/resource-center-profile-page/resource-center-profile-page'
import {
  getPartners,
  getPartnerOffices
} from '../../../../../../src/client/services/resource-center-lookup.js'
import { submitProfile } from '../../../../../../src/client/actions/resource-center-profile'
jest.mock('../../../../../../src/client/services/resource-center-lookup.js')
jest.mock('../../../../../../src/client/actions/resource-center-profile')

const idPrefix = 'resource-center-profile-'
const partners = {
  partner1: [
    {
      name1: 'New York Small Business Development Center ~ State University of New York (SUNY) System',
      name2: 'New York State Small Business Development Center ~ Pace University',
      street1: '163 William Street',
      street2: '16th Floor',
      city: 'New York',
      state: 'NY',
      zip: '10038',
      phone: ''
    }
  ],
  partner2: [
    {
      name1: 'Colorado Small Business Development Center ~ State of Colorado',
      name2: 'Colorado Small Business Development Center ~ Boulder Small Business Development Center',
      street1: '1001 Arapahoe Ave',
      street2: '',
      city: 'Boulder',
      state: 'CO',
      zip: '80302',
      phone: '303-442-1475'
    },
    {
      name1: 'Colorado Small Business Development Center ~ State of Colorado',
      name2: 'Colorado Small Business Development Center ~ Pikes Peak Small Business Development Center',
      street1: '1675 W Garden of the Gods Rd',
      street2: "El Paso County Citizen's Center",
      city: 'Colorado Springs',
      state: 'CO',
      zip: '80907',
      phone: ''
    }
  ]
}

getPartners.mockImplementation(() => Object.keys(partners))

getPartnerOffices.mockImplementation(partner => partners[partner])

describe('ResourceCenterProfilePage', () => {
  test('Office select is hidden until partner is selected and address is hidden until office is selected', () => {
    const component = shallow(<ResourceCenterProfilePage />)
    const selectedPartner = getPartners()[0]
    const selectedOffice = getPartnerOffices(selectedPartner)[0]
    let officeSelect = component.find('#' + idPrefix + 'office')
    let shouldUpdateContactInfo = component.find('#' + idPrefix + 'should-update-contact-info')
    expect(officeSelect.node).toBeUndefined()
    expect(shouldUpdateContactInfo.node).toBeUndefined()
    const partnerSelect = component.find('#' + idPrefix + 'partner')
    partnerSelect.simulate('change', { value: selectedPartner })
    officeSelect = component.find('#' + idPrefix + 'office')
    expect(officeSelect.node).toBeDefined()
    shouldUpdateContactInfo = component.find('#' + idPrefix + 'should-update-contact-info')
    expect(shouldUpdateContactInfo.node).toBeUndefined()
    officeSelect.simulate('change', { value: selectedOffice })
    shouldUpdateContactInfo = component.find('#' + idPrefix + 'should-update-contact-info')
    expect(shouldUpdateContactInfo.node).toBeDefined()
  })

  test('Entering non-optional fields allows form to be submitted', () => {
    const component = shallow(<ResourceCenterProfilePage />)
    const selectedPartner = getPartners()[1]
    const selectedOffice = getPartnerOffices(selectedPartner)[1]
    const partnerSelect = component.find('#' + idPrefix + 'partner')
    partnerSelect.simulate('change', { value: selectedPartner })
    const officeSelect = component.find('#' + idPrefix + 'office')
    officeSelect.simulate('change', { value: selectedOffice })
    const shouldUpdateContactInfo = component.find('#' + idPrefix + 'should-update-contact-info')
    shouldUpdateContactInfo.simulate('change', true)
    const url = component.find('#' + idPrefix + 'url')
    url.simulate('change', { target: { name: 'url', value: 'http://www.google.com' } })
    const serviceArea = component.find('#' + idPrefix + 'service-area')
    serviceArea.simulate('change', { value: 'Entire Nation' })
    const expertiseCheckbox = component.find('#' + idPrefix + 'Networking')
    //not sure why this is not working, manually calling for now
    expertiseCheckbox.simulate('change', { target: { checked: true, name: 'Networking' } })
    component.instance().handleCheckbox({ target: { checked: true, name: 'Networking' } }, 'expertise')

    const businessStage = component.find('#' + idPrefix + 'business-stage')
    businessStage.simulate('change', 'New and early stage businesses')
    const servicesCheckbox = component.find('#' + idPrefix + 'Appointments')
    // not sure why this is not working, manually calling for now
    servicesCheckbox.simulate('change', { target: { checked: true, name: 'Appointments' } })
    component.instance().handleCheckbox({ target: { checked: true, name: 'Appointments' } }, 'services')

    const submit = component.find('#' + idPrefix + 'form-submit-button')
    submit.simulate('click')
    const errorMsg = component.find('FormErrorMessage')
    expect(errorMsg.node).toBeUndefined()
    expect(component.instance().state.profile).toMatchSnapshot()
  })

  test('Missing required fields causes error to show', () => {
    const component = shallow(<ResourceCenterProfilePage />)

    const submit = component.find('#' + idPrefix + 'form-submit-button').parent()
    submit.simulate('click', { preventDefault: () => {} })
    const errorMsg = component.find('FormErrorMessage')
    expect(errorMsg.node).toBeDefined()
    expect(errorMsg.node).toMatchSnapshot()
  })

  test('Getting isSubmitted=true response causes message to display', () => {
    const component = shallow(<ResourceCenterProfilePage isSubmitComplete={true} />)
    const submissionMsg = component.find('#' + idPrefix + 'submission-complete')
    expect(submissionMsg.node).toBeDefined()
    expect(submissionMsg.node).toMatchSnapshot()
  })
})
