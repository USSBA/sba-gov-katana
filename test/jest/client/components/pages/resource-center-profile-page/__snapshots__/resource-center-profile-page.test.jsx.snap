// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceCenterProfilePage Entering non-optional fields allows form to be submitted 1`] = `
Object {
  "address": "1675 W Garden of the Gods Rd, El Paso County Citizen's Center, Colorado Springs, CO, 80907",
  "businessStage": "New and early stage businesses",
  "expertise": Array [
    "Networking",
  ],
  "hours": Object {
    "fridayClose": null,
    "fridayOpen": null,
    "mondayClose": null,
    "mondayOpen": null,
    "saturdayClose": null,
    "saturdayOpen": null,
    "sundayClose": null,
    "sundayOpen": null,
    "thursdayClose": null,
    "thursdayOpen": null,
    "tuesdayClose": null,
    "tuesdayOpen": null,
    "wednesdayClose": null,
    "wednesdayOpen": null,
  },
  "languages": Array [],
  "name": "Colorado Small Business Development Center ~ State of Colorado | Colorado Small Business Development Center ~ Pikes Peak Small Business Development Center",
  "needsUpdating": false,
  "phone": "",
  "serviceArea": "Entire Nation",
  "services": Array [
    "Appointments",
  ],
  "type": "partner2",
  "url": "http://www.google.com",
}
`;

exports[`ResourceCenterProfilePage Getting isSubmitted=true response causes message to display 1`] = `
<div
  id="resource-center-profile-submission-complete"
>
  <h1>
    One Last Step
  </h1>
  <p>
    Thanks for completing this form. Your submission will help us better match you with your ideal clients.
  </p>
  <p>
    To update your office address and phone number, contact SBA's
     
    <a
      href="mailto:edmis@sba.gov?subject=Resource Center Profile- updates"
    >
      Office of Entrepreneurial Development
    </a>
  </p>
</div>
`;

exports[`ResourceCenterProfilePage Missing required fields causes error to show 1`] = `
<FormErrorMessage
  errorText="Please answer all required questions before submitting."
/>
`;
